go/login/login.go

```diff
@@ -0,0 +1,388 @@
+package login
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/google/glome/go/glome"
+)
+
+const (
+	// Minimal acceptable length of a handshake. 1 byte for the prefix, 32 bytes for the key
+	minHandshakeLen = 1 + glome.PublicKeySize
+)
+
+var (
+	// ErrInvalidURLFormat denotes that the URL has a wrong format.
+	ErrInvalidURLFormat = fmt.Errorf("URL is malformed")
+	// ErrInvalidHandshakeLen denotes that the handshake is too short.
+	ErrInvalidHandshakeLen = fmt.Errorf("handshake length is small: should be at least %d", minHandshakeLen)
+	// ErrVersionNotSupported denotes that the version of glome-login URL format is not supported.
+	ErrVersionNotSupported = fmt.Errorf("version not supported")
+	// ErrInvalidPrefixType denotes that the prefix type is invalid.
+	ErrInvalidPrefixType = fmt.Errorf("invalid prefix type")
+	// ErrIncorrectTag denotes that received tag is incorrect.
+	ErrIncorrectTag = fmt.Errorf("invalid tag")
+	// ErrResponseNotInitialized denotes that the response is not initialized.
+	ErrResponseNotInitialized = fmt.Errorf("response is not initialized")
+	// ErrServerKeyNotFound server key not found.
+	ErrServerKeyNotFound = fmt.Errorf("server key not found")
+)
+
+var (
+	validURLPrefix = regexp.MustCompile(`/(?P<v>v[1-9][0-9]*)/(?P<handshake>[\w=-]+)/`)
+)
+
+// Message struct represents the context required for authorization.
+// It contains HostIDType - type of identity, HostID - identity of the target (e.g. hostname, serial number, etc.),
+// Action - action that is being authorized.
+type Message struct {
+	HostIDType string
+	HostID     string
+	Action     string
+}
+
+// Construct returns a message from a Message according to the format: [<hostid-type>:]<hostid>[/<action>].
+// URL escaping is optional.
+func (m *Message) Construct(esc bool) []byte {
+	hostIDType := ""
+	hostID := ""
+	action := ""
+	if esc {
+		hostIDType = escape(m.HostIDType)
+		hostID = escape(m.HostID)
+	} else {
+		hostIDType = m.HostIDType
+		hostID = m.HostID
+	}
+
+	if hostIDType != "" {
+		hostIDType += ":"
+	}
+
+	if m.Action != "" {
+		action = "/" + m.Action
+	}
+	return []byte(hostIDType + hostID + action)
+}
+
+// Escapes the string so it can be safely placed inside a URL path segment,
+// replacing "/#?" special characters and not replacing "!*'();:@&=+$,[]" special characters.
+func escape(s string) string {
+	res := url.PathEscape(s)
+	for _, c := range "!*'();:@&=+$,[]" {
+		st := string(c)
+		strings.Replace(res, url.PathEscape(st), st, -1)
+	}
+	return res
+}
+
+// Handshake struct represents the context required for constructing the handshake.
+// It contains Prefix - either service key or its id, UserKey - user's public ephemeral key,
+// MessageTagPrefix - tag calculated under Message.
+type Handshake struct {
+	Prefix           byte
+	UserKey          glome.PublicKey
+	MessageTagPrefix []byte
+}
+
+// URLResponse struct represents the context required for the URL constructing.
```

Comment:
```suggestion
// URLResponse represents the context required for the construction of the URL.
```

Also same: move struct member comments to the individual members.

Reply:
Will do!